src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java

```diff
@@ -0,0 +1,173 @@
+import org.pac4j.saml.client.SAML2ClientConfiguration;
+/*
+import sun.security.tools.keytool.CertAndKeyGen;
+import sun.security.x509.X500Name;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.security.KeyPairGenerator;
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.SecureRandom;
+import java.security.cert.X509Certificate;
+import java.util.Date;
+*/
+import java.util.logging.Logger;
+
+import static java.util.logging.Level.*;
+import static org.opensaml.saml.common.xml.SAMLConstants.SAML2_REDIRECT_BINDING_URI;
+
+/**
+ * Overall wrapper to all operation using OpenSAML library, this allow to load the Service Loaders properly
+ * <p>
+ * https://wiki.shibboleth.net/confluence/display/OS30/Initialization+and+Configuration
+ * http://docs.oracle.com/javase/6/docs/api/java/util/ServiceLoader.html
+ * https://stackoverflow.com/questions/37948303/opensaml3-resource-not-found-default-config-xml-in-osgi-container/39004323#39004323
+ *
+ * @param <T> process return type
+ */
+public abstract class OpenSAMLWrapper<T> {
+    private static final Logger LOG = Logger.getLogger(OpenSAMLWrapper.class.getName());
+    private static final BundleKeyStore KS = new BundleKeyStore();
+
+    protected SamlPluginConfig samlPluginConfig;
+    protected StaplerRequest request;
+    protected StaplerResponse response;
+
+    /**
+     * Initialize the OpenSaml services and run the process defined on the abstract method process().
+     *
+     * @return process return object
+     */
+    public T get() {
+        T ret = null;
+        try {
+            LOG.finest("adapt TCCL");
+            Thread thread = Thread.currentThread();
+            ClassLoader loader = thread.getContextClassLoader();
+            thread.setContextClassLoader(InitializationService.class.getClassLoader());
+            try {
+                InitializationService.initialize();
+                ret = process();
+            } finally {
+                LOG.finest("reset TCCL");
+                thread.setContextClassLoader(loader);
+            }
+        } catch (InitializationException e) {
+            //FIXME [kuisathaverat] throw exception.
+            LOG.log(SEVERE, "Could not initialize opensaml service.", e);
+            throw new IllegalStateException(e);
+        }
+        return ret;
+    }
+
+    /**
+     * Process to run using the OpenSaml services.
+     *
+     * @return process return type
+     */
+    abstract T process();
+
+    /**
+     * @return J2E Context from the current HTTP request and response.
+     */
+    protected WebContext createWebContext() {
+        return new J2EContext(request, response);
+    }
+
+
+    /**
+     * @return a SAML2Client object to interact with the IdP service.
+     */
+    protected SAML2Client createSAML2Client() {
+        //FIXME [kuisathaverat] Throws NPE I do not like it
+        Preconditions.checkNotNull(samlPluginConfig.getIdpMetadata());
```

Comment:
If the client is not usable without the IdP metadata, then just make it required (and add a validation check) in the configuration and assume it is non-null anywhere else.